src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray.cs

```diff
@@ -395,6 +395,19 @@ public static ImmutableArray<TSource> ToImmutableArray<TSource>(this IEnumerable
             return CreateRange(items);
         }
 
+        /// <summary>
+        /// Returns an immutable copy of the current contents of the builder's collection.
+        /// </summary>
+        /// <param name="builder">The builder to create the immutable array from.</param>
+        /// <returns>An immutable array.</returns>
+        [Pure]
```

Comment:
Given that `[Pure]` has no runtime impact, why isn't the attribute included in the ref assemblies so that analyzers can do their work based on it?

Reply:
Looking at it (and not the author) I'd guess that it's because none of the other definitions in the ref assembly have them and he followed the existing pattern on the assumption that someone in the past had good reason to do this.

Adding them in ref would require a dependency on `System.Diagnostics.Contracts` which would seem an odd thing to do.

Reply:
Agreed: this PR itself looks consistent with precedent. So my question was perhaps more directed at the authors of the ref assemblies in general. @danmosemsft @csharpfritz perhaps?

Reply:
I think I was included incorrectly on this...

Reply:
`[Pure]` is part of Contracts, right? We don't use or enforce those and generally have been removing contract annotations. I believe we discussed this before in the context of immutable collections in fact and @AArnott did you indicate you didn't believe in them either? In which case I guess these can all be removed unless they are somehow still useful to .NET Framework consumers even though they're not on the ref assembly and probably not consistent either.

Reply:
@danmosemsft  I think `PureAttribute` preceded Code Contracts. And ya, I don't use Code Contracts at all any more nor hear of anyone else using them.
But `[Pure]` is useful because one can write analyzers that will flag a warning when a `[Pure]` method is executed without doing anything with the result. Particularly on the immutable collections that's relevant since someone may simply right:

```cs
set.Add(a);
```

And be confused when `set` doesn't contain `a`. But if `Add` were decorated with `[Pure]` an analyzer could say "hey, aren't you supposed to do something with the return value?" which would guide the user to fixing it to be: `set = set.Add(a);`.

So I am in favor of correctly applying `[Pure]` in many places, *especially* ref assemblies since that's what analyzers would have visibility into.

Reply:
From a quick search we don't have any `[Pure]` attributes in ref contracts so far in corefx - but we do have it in several places in collections: https://github.com/dotnet/corefx/search?p=3&q=Pure&unscoped_q=Pure

So it doesn't seem we use them at all, or enforce them to be applied in the ref contracts.

Reply:
I suspect .NET doesn't define many pure methods anyway (regardless whether they are tagged as `[Pure]`). Immutable types are a relatively new area and the collections are a significant step in that direction, so I'd love to see this attribute thoroughly applied to that library, at least. Would you accept a PR? 
Maybe I'd even combine it with an analyzer assembly that the immutable collections package depends on that will scan for Pure calls as I mentioned.

Reply:
Would it go on eg `String.Replace`, etc? 

@jkotas is this something we want to start doing? It seems to me we likely want an analyzer proving it adds value first. @jaredpar is this something that CSC might start using (eg built in analyzer to spot dropped return codes)

Reply:
> It seems to me we likely want an analyzer proving it adds value first. 

Right. Based on the past experiments with Code Contracts, it is very hard to get a return on investment with this. The annotations do catch a few bugs, but they require a prohibitive amount of manual work to do that.

Reply:
> Based on the past experiments with Code Contracts, it is very hard to get a return on investment with this.

Code Contracts preceded roslyn analyzers, and required custom build steps, VS extensions, and/or post-compilation IL-rewriting . Roslyn analyzers are proven to be effective and popular. Let's not let Code Contracts' failure deter us from tapping Analyzers for solving this simple issue.

Reply:
The Pure attribute is checked by Rule CA1806 in Microsoft.CodeQuality.Analyzers.

Reply:
Thanks for redrawing attention to this, @Grauenwolf. I've filed a separate issue to track the issue: https://github.com/dotnet/corefx/issues/34778